clipboard file: C2S transfer is not working with Thunar but not working  with Nautilus 3.38

## Overview

Environment is the same as #1992. I think some clipboard events that xrdp-chansrv generate are not compatible with the latest Nautilus.

![スクリーンショット 2021-09-09 16 34 45](https://user-images.githubusercontent.com/941609/132646732-99fe670b-4fa7-4289-affa-2b9459934b32.png)

## C2S transfer is working with Thunar but not working with Nautilus 

https://user-images.githubusercontent.com/941609/132649704-36be569e-6828-478a-84b6-462fe4418ad1.mov

## S2C transfer is working  both with Nautilus and Thunar 

https://user-images.githubusercontent.com/941609/132646086-67f01358-bbce-4172-99b8-2967b03a7ffe.mov

<details>

<summary>
chansrv log pasting in Nautilus
</summary>

```
Nautilus

[20210909-08:30:31] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:30:31] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:30:31] [DEBUG] [my_trans_data_in(chansrv.c:909)] my_trans_data_in:
[20210909-08:30:31] [DEBUG] [process_message_channel_data(chansrv.c:446)] process_message_channel_data: chan_id 2 chan_flags 19
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1523)] clipboard_data_in: chan_id 2 chan_flags 0x13 length 84 total_length 84 in_request 0 g_ins->size 8192
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1586)] clipboard_data_in: clip_msg_id 2 clip_msg_status 0 clip_msg_len 76
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1591)] clipboard_data_in: 2
[20210909-08:30:31] [DEBUG] [clipboard_process_format_announce(clipboard.c:985)] clipboard_process_format_announce: CLIPRDR_FORMAT_ANNOUNCE
[20210909-08:30:31] [DEBUG] [clipboard_process_format_announce(clipboard.c:987)] clipboard_process_format_announce 76
[20210909-08:30:31] [DEBUG] [clipboard_send_format_ack(clipboard.c:542)] clipboard_send_format_ack: data out, sending CLIPRDR_FORMAT_ACK (clip_msg_id = 3)
[20210909-08:30:31] [DEBUG] [clipboard_process_format_announce(clipboard.c:1017)] clipboard_process_format_announce: formatId 0x0000c0bc wszFormatName [FileGroupDescriptorW] clip_msg_len 30
[20210909-08:30:31] [DEBUG] [clipboard_process_format_announce(clipboard.c:1017)] clipboard_process_format_announce: formatId 0x0000c0ba wszFormatName [FileContents] clip_msg_len 0
[20210909-08:30:31] [DEBUG] [clipboard_set_selection_owner(clipboard.c:851)] clipboard_set_selection_owner:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2478)] clipboard_xevent: got XFixesSetSelectionOwnerNotify
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_owner_notify(clipboard.c:1665)] clipboard_event_selection_owner_notify: 0x400003
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_owner_notify(clipboard.c:1666)] clipboard_event_selection_owner_notify: window 4194307 subtype 0 owner 4194307 g_wnd 4194307
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_owner_notify(clipboard.c:1672)] clipboard_event_selection_owner_notify: matches g_wnd
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_owner_notify(clipboard.c:1673)] clipboard_event_selection_owner_notify: skipping, owner == g_wnd
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 35651595 .owner 4194307 .selection 246 'CLIPBOARD' .target 250 .property 296
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2112)] clipboard_event_selection_request: g_targets_atom
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2114)] clipboard_event_selection_request: g_targets_atom
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2130)]   reporting text/uri-list
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2133)]   reporting x-special/gnome-copied-files
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2138)]   reporting 7 formats
[20210909-08:30:31] [DEBUG] [clipboard_provide_selection(clipboard.c:932)] clipboard_provide_selection: bytes 56
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 39846149 .owner 4194307 .selection 246 'CLIPBOARD' .target 250 .property 296
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2112)] clipboard_event_selection_request: g_targets_atom
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2114)] clipboard_event_selection_request: g_targets_atom
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2130)]   reporting text/uri-list
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2133)]   reporting x-special/gnome-copied-files
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2138)]   reporting 7 formats
[20210909-08:30:31] [DEBUG] [clipboard_provide_selection(clipboard.c:932)] clipboard_provide_selection: bytes 56
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 44040194 .owner 4194307 .selection 246 'CLIPBOARD' .target 250 .property 296
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2112)] clipboard_event_selection_request: g_targets_atom
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2114)] clipboard_event_selection_request: g_targets_atom
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2130)]   reporting text/uri-list
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2133)]   reporting x-special/gnome-copied-files
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2138)]   reporting 7 formats
[20210909-08:30:31] [DEBUG] [clipboard_provide_selection(clipboard.c:932)] clipboard_provide_selection: bytes 56
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46138398 .owner 4194307 .selection 246 'CLIPBOARD' .target 440 .property 296
[20210909-08:30:31] [ERROR] [clipboard_event_selection_request(clipboard.c:2228)] clipboard_event_selection_request: unknown target text/plain;charset=utf-8
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46289069 .owner 4194307 .selection 246 'CLIPBOARD' .target 440 .property 296
[20210909-08:30:31] [ERROR] [clipboard_event_selection_request(clipboard.c:2228)] clipboard_event_selection_request: unknown target text/plain;charset=utf-8
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46289072 .owner 4194307 .selection 246 'CLIPBOARD' .target 440 .property 296
[20210909-08:30:31] [ERROR] [clipboard_event_selection_request(clipboard.c:2228)] clipboard_event_selection_request: unknown target text/plain;charset=utf-8
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46138398 .owner 4194307 .selection 246 'CLIPBOARD' .target 243 .property 296
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:507)] clipboard_send_data_request:
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:508)] clipboard_send_data_request: 13
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:519)] clipboard_send_data_request: data out, sending CLIPRDR_DATA_REQUEST (clip_msg_id = 4)
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46289075 .owner 4194307 .selection 246 'CLIPBOARD' .target 243 .property 296
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:507)] clipboard_send_data_request:
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:508)] clipboard_send_data_request: 13
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:519)] clipboard_send_data_request: data out, sending CLIPRDR_DATA_REQUEST (clip_msg_id = 4)
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:30:31] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46289078 .owner 4194307 .selection 246 'CLIPBOARD' .target 243 .property 296
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:507)] clipboard_send_data_request:
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:508)] clipboard_send_data_request: 13
[20210909-08:30:31] [DEBUG] [clipboard_send_data_request(clipboard.c:519)] clipboard_send_data_request: data out, sending CLIPRDR_DATA_REQUEST (clip_msg_id = 4)
[20210909-08:30:31] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:30:31] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
[20210909-08:30:31] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:30:31] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:30:31] [DEBUG] [my_trans_data_in(chansrv.c:909)] my_trans_data_in:
[20210909-08:30:31] [DEBUG] [process_message_channel_data(chansrv.c:446)] process_message_channel_data: chan_id 2 chan_flags 19
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1523)] clipboard_data_in: chan_id 2 chan_flags 0x13 length 60 total_length 60 in_request 1 g_ins->size 8192
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1586)] clipboard_data_in: clip_msg_id 5 clip_msg_status 1 clip_msg_len 52
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1591)] clipboard_data_in: 5
[20210909-08:30:31] [DEBUG] [clipboard_process_data_response(clipboard.c:1219)] clipboard_process_data_response:
[20210909-08:30:31] [DEBUG] [clipboard_process_data_response(clipboard.c:1252)] clipboard_process_data_response: CLIPRDR_DATA_RESPONSE
[20210909-08:30:31] [DEBUG] [clipboard_provide_selection_c2s(clipboard.c:873)] clipboard_provide_selection_c2s: bytes 25
[20210909-08:30:31] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
[20210909-08:30:31] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:30:31] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:30:31] [DEBUG] [my_trans_data_in(chansrv.c:909)] my_trans_data_in:
[20210909-08:30:31] [DEBUG] [process_message_channel_data(chansrv.c:446)] process_message_channel_data: chan_id 2 chan_flags 19
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1523)] clipboard_data_in: chan_id 2 chan_flags 0x13 length 60 total_length 60 in_request 0 g_ins->size 8192
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1586)] clipboard_data_in: clip_msg_id 5 clip_msg_status 1 clip_msg_len 52
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1591)] clipboard_data_in: 5
[20210909-08:30:31] [DEBUG] [clipboard_process_data_response(clipboard.c:1219)] clipboard_process_data_response:
[20210909-08:30:31] [DEBUG] [clipboard_process_data_response(clipboard.c:1252)] clipboard_process_data_response: CLIPRDR_DATA_RESPONSE
[20210909-08:30:31] [DEBUG] [clipboard_provide_selection_c2s(clipboard.c:873)] clipboard_provide_selection_c2s: bytes 25
[20210909-08:30:31] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
[20210909-08:30:31] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:30:31] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:30:31] [DEBUG] [my_trans_data_in(chansrv.c:909)] my_trans_data_in:
[20210909-08:30:31] [DEBUG] [process_message_channel_data(chansrv.c:446)] process_message_channel_data: chan_id 2 chan_flags 19
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1523)] clipboard_data_in: chan_id 2 chan_flags 0x13 length 60 total_length 60 in_request 0 g_ins->size 8192
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1586)] clipboard_data_in: clip_msg_id 5 clip_msg_status 1 clip_msg_len 52
[20210909-08:30:31] [DEBUG] [clipboard_data_in(clipboard.c:1591)] clipboard_data_in: 5
[20210909-08:30:31] [DEBUG] [clipboard_process_data_response(clipboard.c:1219)] clipboard_process_data_response:
[20210909-08:30:31] [DEBUG] [clipboard_process_data_response(clipboard.c:1252)] clipboard_process_data_response: CLIPRDR_DATA_RESPONSE
[20210909-08:30:31] [DEBUG] [clipboard_provide_selection_c2s(clipboard.c:873)] clipboard_provide_selection_c2s: bytes 25
[20210909-08:30:31] [ERROR] [xcommon_error_handler(xcommon.c:59)] X error [BadWindow (invalid Window parameter)](3) opcodes 18/0 resource 0x2c250b6
[20210909-08:30:31] [ERROR] [xcommon_error_handler(xcommon.c:59)] X error [BadWindow (invalid Window parameter)](3) opcodes 25/0 resource 0x2c250b6
[20210909-08:30:31] [ERROR] [xcommon_error_handler(xcommon.c:59)] X error [BadWindow (invalid Window parameter)](3) opcodes 18/0 resource 0x2c250b6
[20210909-08:30:31] [ERROR] [xcommon_error_handler(xcommon.c:59)] X error [BadWindow (invalid Window parameter)](3) opcodes 25/0 resource 0x2c250b6
[20210909-08:30:31] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
```

</details>



<details>
<summary>
chansrv logs pasting in Thunar
</summary>

```
Thunar

[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:34:00] [DEBUG] [my_trans_data_in(chansrv.c:909)] my_trans_data_in:
[20210909-08:34:00] [DEBUG] [process_message_channel_data(chansrv.c:446)] process_message_channel_data: chan_id 2 chan_flags 19
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1523)] clipboard_data_in: chan_id 2 chan_flags 0x13 length 84 total_length 84 in_request 0 g_ins->size 30472
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1586)] clipboard_data_in: clip_msg_id 2 clip_msg_status 0 clip_msg_len 76
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1591)] clipboard_data_in: 2
[20210909-08:34:00] [DEBUG] [clipboard_process_format_announce(clipboard.c:985)] clipboard_process_format_announce: CLIPRDR_FORMAT_ANNOUNCE
[20210909-08:34:00] [DEBUG] [clipboard_process_format_announce(clipboard.c:987)] clipboard_process_format_announce 76
[20210909-08:34:00] [DEBUG] [clipboard_send_format_ack(clipboard.c:542)] clipboard_send_format_ack: data out, sending CLIPRDR_FORMAT_ACK (clip_msg_id = 3)
[20210909-08:34:00] [DEBUG] [clipboard_process_format_announce(clipboard.c:1017)] clipboard_process_format_announce: formatId 0x0000c0bc wszFormatName [FileGroupDescriptorW] clip_msg_len 30
[20210909-08:34:00] [DEBUG] [clipboard_process_format_announce(clipboard.c:1017)] clipboard_process_format_announce: formatId 0x0000c0ba wszFormatName [FileContents] clip_msg_len 0
[20210909-08:34:00] [DEBUG] [clipboard_set_selection_owner(clipboard.c:851)] clipboard_set_selection_owner:
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2478)] clipboard_xevent: got XFixesSetSelectionOwnerNotify
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_owner_notify(clipboard.c:1665)] clipboard_event_selection_owner_notify: 0x400003
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_owner_notify(clipboard.c:1666)] clipboard_event_selection_owner_notify: window 4194307 subtype 0 owner 4194307 g_wnd 4194307
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_owner_notify(clipboard.c:1672)] clipboard_event_selection_owner_notify: matches g_wnd
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_owner_notify(clipboard.c:1673)] clipboard_event_selection_owner_notify: skipping, owner == g_wnd
[20210909-08:34:00] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 44040194 .owner 4194307 .selection 246 'CLIPBOARD' .target 250 .property 296
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2112)] clipboard_event_selection_request: g_targets_atom
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2114)] clipboard_event_selection_request: g_targets_atom
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2130)]   reporting text/uri-list
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2133)]   reporting x-special/gnome-copied-files
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2138)]   reporting 7 formats
[20210909-08:34:00] [DEBUG] [clipboard_provide_selection(clipboard.c:932)] clipboard_provide_selection: bytes 56
[20210909-08:34:00] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 39846149 .owner 4194307 .selection 246 'CLIPBOARD' .target 250 .property 296
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2112)] clipboard_event_selection_request: g_targets_atom
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2114)] clipboard_event_selection_request: g_targets_atom
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2130)]   reporting text/uri-list
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2133)]   reporting x-special/gnome-copied-files
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2138)]   reporting 7 formats
[20210909-08:34:00] [DEBUG] [clipboard_provide_selection(clipboard.c:932)] clipboard_provide_selection: bytes 56
[20210909-08:34:00] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 35651595 .owner 4194307 .selection 246 'CLIPBOARD' .target 250 .property 296
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2112)] clipboard_event_selection_request: g_targets_atom
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2114)] clipboard_event_selection_request: g_targets_atom
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2130)]   reporting text/uri-list
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2133)]   reporting x-special/gnome-copied-files
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2138)]   reporting 7 formats
[20210909-08:34:00] [DEBUG] [clipboard_provide_selection(clipboard.c:932)] clipboard_provide_selection: bytes 56
[20210909-08:34:00] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46138398 .owner 4194307 .selection 246 'CLIPBOARD' .target 440 .property 296
[20210909-08:34:00] [ERROR] [clipboard_event_selection_request(clipboard.c:2228)] clipboard_event_selection_request: unknown target text/plain;charset=utf-8
[20210909-08:34:00] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46302839 .owner 4194307 .selection 246 'CLIPBOARD' .target 440 .property 296
[20210909-08:34:00] [ERROR] [clipboard_event_selection_request(clipboard.c:2228)] clipboard_event_selection_request: unknown target text/plain;charset=utf-8
[20210909-08:34:00] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:34:00] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46138398 .owner 4194307 .selection 246 'CLIPBOARD' .target 243 .property 296
[20210909-08:34:00] [DEBUG] [clipboard_send_data_request(clipboard.c:507)] clipboard_send_data_request:
[20210909-08:34:00] [DEBUG] [clipboard_send_data_request(clipboard.c:508)] clipboard_send_data_request: 13
[20210909-08:34:00] [DEBUG] [clipboard_send_data_request(clipboard.c:519)] clipboard_send_data_request: data out, sending CLIPRDR_DATA_REQUEST (clip_msg_id = 4)
[20210909-08:34:00] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:34:00] [DEBUG] [clipboard_xevent(clipboard.c:2434)] clipboard_xevent: event detected
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2098)] clipboard_event_selection_request: 0x128
[20210909-08:34:00] [DEBUG] [clipboard_event_selection_request(clipboard.c:2099)] clipboard_event_selection_request: g_wnd 4194307, .requestor 46302866 .owner 4194307 .selection 246 'CLIPBOARD' .target 243 .property 296
[20210909-08:34:00] [DEBUG] [clipboard_send_data_request(clipboard.c:507)] clipboard_send_data_request:
[20210909-08:34:00] [DEBUG] [clipboard_send_data_request(clipboard.c:508)] clipboard_send_data_request: 13
[20210909-08:34:00] [DEBUG] [clipboard_send_data_request(clipboard.c:519)] clipboard_send_data_request: data out, sending CLIPRDR_DATA_REQUEST (clip_msg_id = 4)
[20210909-08:34:00] [DEBUG] [rail_xevent(rail.c:1803)] chansrv::rail_xevent:
[20210909-08:34:00] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:34:00] [DEBUG] [my_trans_data_in(chansrv.c:909)] my_trans_data_in:
[20210909-08:34:00] [DEBUG] [process_message_channel_data(chansrv.c:446)] process_message_channel_data: chan_id 2 chan_flags 19
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1523)] clipboard_data_in: chan_id 2 chan_flags 0x13 length 40 total_length 40 in_request 1 g_ins->size 30472
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1586)] clipboard_data_in: clip_msg_id 5 clip_msg_status 1 clip_msg_len 32
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1591)] clipboard_data_in: 5
[20210909-08:34:00] [DEBUG] [clipboard_process_data_response(clipboard.c:1219)] clipboard_process_data_response:
[20210909-08:34:00] [DEBUG] [clipboard_process_data_response(clipboard.c:1252)] clipboard_process_data_response: CLIPRDR_DATA_RESPONSE
[20210909-08:34:00] [DEBUG] [clipboard_provide_selection_c2s(clipboard.c:873)] clipboard_provide_selection_c2s: bytes 15
[20210909-08:34:00] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:34:00] [DEBUG] [my_trans_data_in(chansrv.c:909)] my_trans_data_in:
[20210909-08:34:00] [DEBUG] [process_message_channel_data(chansrv.c:446)] process_message_channel_data: chan_id 2 chan_flags 19
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1523)] clipboard_data_in: chan_id 2 chan_flags 0x13 length 40 total_length 40 in_request 0 g_ins->size 30472
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1586)] clipboard_data_in: clip_msg_id 5 clip_msg_status 1 clip_msg_len 32
[20210909-08:34:00] [DEBUG] [clipboard_data_in(clipboard.c:1591)] clipboard_data_in: 5
[20210909-08:34:00] [DEBUG] [clipboard_process_data_response(clipboard.c:1219)] clipboard_process_data_response:
[20210909-08:34:00] [DEBUG] [clipboard_process_data_response(clipboard.c:1252)] clipboard_process_data_response: CLIPRDR_DATA_RESPONSE
[20210909-08:34:00] [DEBUG] [clipboard_provide_selection_c2s(clipboard.c:873)] clipboard_provide_selection_c2s: bytes 15
[20210909-08:34:00] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:201)] check_timeout:
[20210909-08:34:00] [DEBUG] [check_timeout(chansrv.c:241)]   count 0
[20210909-08:34:00] [ERROR] [xcommon_error_handler(xcommon.c:59)] X error [BadWindow (invalid Window parameter)](3) opcodes 18/0 resource 0x2c28692
[20210909-08:34:00] [ERROR] [xcommon_error_handler(xcommon.c:59)] X error [BadWindow (invalid Window parameter)](3) opcodes 25/0 resource 0x2c28692
[20210909-08:34:00] [DEBUG] [get_timeout(chansrv.c:153)] get_timeout:
```
</details>

I can reproduce some of this one (Nautilus not working) using the [files.zip](https://github.com/neutrinolabs/xrdp/files/7132609/files.zip) from #1992. I also get this message:-

```
[ERROR] clipboard_event_selection_request: unknown target text/plain;charset=utf-8
```

It's not clear to me whether this actually is an error or not. Provided the application is also requesting `UTF8_STRING`,  in the message to `clipboard_event_selection_notify()`, everything should work OK.

I'll get Thunar loaded on a test machine.

Note: Nemo is a fork of GNOME2 Nautilus. 

I captured chansrv log like `sesman/chansrv/clipboard-notes.txt`. When I copied a file to clipboard, there are the following differences I get.

So chansrv need to put clipboard in new Nautilus compatible format to allow pasting in new Nautilus. But we need to keep compatibility with other file managers like Thunar/Nemo. I'm not sure it is possible to support both.

@jsorg71 I found you committed `clipboard-notes.txt` 9 years ago. Any ideas?
## Nautilus 3.38.2
```
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1819)] clipboard_event_selection_notify:
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1842)] clipboard_event_selection_notify: wnd 0xa00001 prop XRDP_CLIP_PROPERTY_ATOM
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xf9 TIMESTAMP 0x1f
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xf9 TIMESTAMP
[20210910-09:00:58] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xf9
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfa TARGETS 0x1f
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfa TARGETS
[20210910-09:00:58] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xfa
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfb MULTIPLE 0x1f
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfb MULTIPLE
[20210910-09:00:58] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xfb
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfd text/uri-list 0x1f
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfd text/uri-list
[20210910-09:00:58] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1908)] clipboard_event_selection_notify: file
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xf3 UTF8_STRING 0x1f
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xf3 UTF8_STRING
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1af COMPOUND_TEXT 0x1f
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1af COMPOUND_TEXT
[20210910-09:00:59] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0x1af
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1c5 TEXT 0x1f
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1c5 TEXT
[20210910-09:00:59] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0x1c5
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1f STRING 0x1f
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1f STRING
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1bb text/plain;charset=utf-8 0x1f
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1bb text/plain;charset=utf-8
[20210910-09:00:59] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0x1bb
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1c6 text/plain 0x1f
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1c6 text/plain
[20210910-09:00:59] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0x1c6
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1819)] clipboard_event_selection_notify:
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1842)] clipboard_event_selection_notify: wnd 0xa00001 prop XRDP_CLIP_PROPERTY_ATOM
[20210910-09:00:59] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1978)] clipboard_event_selection_notify: text/uri-list data_size 40
```

## Nemo
```
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1819)] clipboard_event_selection_notify:
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1842)] clipboard_event_selection_notify: wnd 0xa00001 prop XRDP_CLIP_PROPERTY_ATOM
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xf9 TIMESTAMP 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xf9 TIMESTAMP
[20210910-09:02:10] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xf9
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfa TARGETS 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfa TARGETS
[20210910-09:02:10] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xfa
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfb MULTIPLE 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfb MULTIPLE
[20210910-09:02:10] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xfb
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfe x-special/gnome-copied-files 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfe x-special/gnome-copied-files
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1908)] clipboard_event_selection_notify: file
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfd text/uri-list 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfd text/uri-list
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1908)] clipboard_event_selection_notify: file
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xf3 UTF8_STRING 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xf3 UTF8_STRING
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1af COMPOUND_TEXT 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1af COMPOUND_TEXT
[20210910-09:02:10] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0x1af
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1c5 TEXT 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1c5 TEXT
[20210910-09:02:10] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0x1c5
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1f STRING 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1f STRING
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1bb text/plain;charset=utf-8 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1bb text/plain;charset=utf-8
[20210910-09:02:10] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0x1bb
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0x1c6 text/plain 0x1f
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0x1c6 text/plain
[20210910-09:02:10] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0x1c6
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1819)] clipboard_event_selection_notify:
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1842)] clipboard_event_selection_notify: wnd 0xa00001 prop XRDP_CLIP_PROPERTY_ATOM
[20210910-09:02:10] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1978)] clipboard_event_selection_notify: text/uri-list data_size 40
```

## Thunar
```
[20210910-09:03:17] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1819)] clipboard_event_selection_notify:
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1842)] clipboard_event_selection_notify: wnd 0xa00001 prop XRDP_CLIP_PROPERTY_ATOM
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xf9 TIMESTAMP 0x1f
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xf9 TIMESTAMP
[20210910-09:03:18] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xf9
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfa TARGETS 0x1f
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfa TARGETS
[20210910-09:03:18] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xfa
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfb MULTIPLE 0x1f
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfb MULTIPLE
[20210910-09:03:18] [ERROR] [clipboard_event_selection_notify(clipboard.c:1913)] clipboard_event_selection_notify: unknown atom 0xfb
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfd text/uri-list 0x1f
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfd text/uri-list
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1908)] clipboard_event_selection_notify: file
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xfe x-special/gnome-copied-files 0x1f
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xfe x-special/gnome-copied-files
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1908)] clipboard_event_selection_notify: file
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1889)] clipboard_event_selection_notify: 0xf3 UTF8_STRING 0x1f
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1892)] clipboard_event_selection_notify: 0xf3 UTF8_STRING
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1819)] clipboard_event_selection_notify:
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1842)] clipboard_event_selection_notify: wnd 0xa00001 prop XRDP_CLIP_PROPERTY_ATOM
[20210910-09:03:18] [DEBUG] [clipboard_event_selection_notify(clipboard.c:1993)] clipboard_event_selection_notify: text/uri-list data_size 43
```

--- a/docs/man/sesman.ini.5.in
+++ b/docs/man/sesman.ini.5.in
@@ -290,6 +290,14 @@ features:-
 .P
 -  copying-and-pasting of files
 .RE
+.TP
+\fBUseNautilus3FlistFormat\fR=\fI[false|true]\fR
+Defaults to \fIfalse\fR.
+Set to \fItrue\fR to make file copy-paste compatible with Nautilus from
+GNOME 3 versions later than 3.29.92. Do not use this for any other reason.
+
+This setting will be removed in a later version of xrdp, when GNOME 3 is
+no longer supported.
 
 .SH "SESSIONS VARIABLES"
 All entries in the \fB[SessionVariables]\fR section are set as

--- a/sesman/chansrv/chansrv_config.c
+++ b/sesman/chansrv/chansrv_config.c
@@ -39,7 +39,7 @@
 #define DEFAULT_ENABLE_FUSE_MOUNT           1
 #define DEFAULT_FUSE_MOUNT_NAME             "xrdp-client"
 #define DEFAULT_FILE_UMASK                  077
-
+#define DEFAULT_USE_NAUTILUS3_FLIST_FORMAT  0
 /**
  * Type used for passing a logging function about
  */
@@ -161,7 +161,7 @@ read_config_chansrv(log_func_t logmsg,
         {
             cfg->enable_fuse_mount = g_text2bool(value);
         }
-        if (g_strcasecmp(name, "FuseMountName") == 0)
+        else if (g_strcasecmp(name, "FuseMountName") == 0)
         {
             g_free(cfg->fuse_mount_name);
             cfg->fuse_mount_name = g_strdup(value);
@@ -175,6 +175,10 @@ read_config_chansrv(log_func_t logmsg,
         else if (g_strcasecmp(name, "FileUmask") == 0)
         {
             cfg->file_umask = strtol(value, NULL, 0);
+        }
+        else if (g_strcasecmp(name, "UseNautilus3FlistFormat") == 0)
+        {
+            cfg->use_nautilus3_flist_format = g_text2bool(value);
         }
     }
 
@@ -206,6 +210,7 @@ new_config(void)
         cfg->restrict_outbound_clipboard = DEFAULT_RESTRICT_OUTBOUND_CLIPBOARD;
         cfg->fuse_mount_name = fuse_mount_name;
         cfg->file_umask = DEFAULT_FILE_UMASK;
+        cfg->use_nautilus3_flist_format = DEFAULT_USE_NAUTILUS3_FLIST_FORMAT;
     }
 
     return cfg;
@@ -290,6 +295,8 @@ config_dump(struct config_chansrv *config)
               g_bool2text(config->enable_fuse_mount));
     g_writeln("    FuseMountName:             %s", config->fuse_mount_name);
     g_writeln("    FileMask:                  0%o", config->file_umask);
+    g_writeln("    Nautilus 3 Flist Format:   %s",
+              g_bool2text(config->use_nautilus3_flist_format));
 }
 
 /******************************************************************************/

--- a/sesman/chansrv/chansrv_config.h
+++ b/sesman/chansrv/chansrv_config.h
@@ -36,6 +36,9 @@ struct config_chansrv
     char *fuse_mount_name;
     /** FileUmask from sesman.ini */
     mode_t file_umask;
+
+    /** Whether to use nautilus3-compatible file lists for the clipboard */
+    int use_nautilus3_flist_format;
 };
 
 

--- a/sesman/chansrv/clipboard.c
+++ b/sesman/chansrv/clipboard.c
@@ -1244,7 +1244,28 @@ clipboard_process_data_response_for_file(struct stream *s,
     else if ((g_clip_c2s.type == XA_STRING) ||
              (g_clip_c2s.type == g_utf8_atom))
     {
-        rv = clipboard_c2s_in_files(s, g_clip_c2s.data, flist_size, "");
+        if (g_cfg->use_nautilus3_flist_format)
+        {
+            /*
+             * This file list format is only used by GNOME 3
+             * versions >= 3.29.92. It is not used by GNOME 4. Remove
+             * this workaround when GNOME 3 is no longer supported by
+             * long-term distros */
+#define LIST_PREFIX "x-special/nautilus-clipboard\ncopy\n"
+#define LIST_PREFIX_LEN (sizeof(LIST_PREFIX) - 1)
+            g_strcpy(g_clip_c2s.data, LIST_PREFIX);
+            rv = clipboard_c2s_in_files(s,
+                                        g_clip_c2s.data + LIST_PREFIX_LEN,
+                                        flist_size - LIST_PREFIX_LEN - 1,
+                                        "file://");
+            g_strcat(g_clip_c2s.data, "\n");
+#undef LIST_PREFIX_LEN
+#undef LIST_PREFIX
+        }
+        else
+        {
+            rv = clipboard_c2s_in_files(s, g_clip_c2s.data, flist_size, "");
+        }
     }
     else
     {

--- a/sesman/chansrv/clipboard.c
+++ b/sesman/chansrv/clipboard.c
@@ -1199,6 +1199,67 @@ clipboard_process_data_response_for_image(struct stream *s,
     return 0;
 }
 
+/**************************************************************************//**
+ * Process a CB_FORMAT_DATA_RESPONSE for an X client requesting a file list
+ *
+ * @param s Stream containing CLIPRDR_FILELIST ([MS-RDPECLIP])
+ * @param clip_msg_status msgFlags from Clipboard PDU Header
+ * @param clip_msg_len dataLen from Clipboard PDU Header
+ *
+ * @return Status
+ */
+static int
+clipboard_process_data_response_for_file(struct stream *s,
+        int clip_msg_status,
+        int clip_msg_len)
+{
+    XSelectionRequestEvent *lxev;
+    int rv = 0;
+
+    LOG_DEVEL(LOG_LEVEL_TRACE, "clipboard_process_data_response_for_file: ");
+    lxev = &g_saved_selection_req_event;
+
+    const int flist_size = 1024 * 1024;
+    g_free(g_clip_c2s.data);
+    g_clip_c2s.data = (char *)g_malloc(flist_size, 0);
+    if (g_clip_c2s.data == NULL)
+    {
+        LOG(LOG_LEVEL_ERROR, "clipboard_process_data_response_for_file: "
+            "Can't allocate memory");
+        rv = 1;
+    }
+    /* text/uri-list */
+    else if (g_clip_c2s.type == g_file_atom1)
+    {
+        rv = clipboard_c2s_in_files(s, g_clip_c2s.data, flist_size);
+    }
+    /* x-special/gnome-copied-files */
+    else if (g_clip_c2s.type == g_file_atom2)
+    {
+        g_strcpy(g_clip_c2s.data, "copy\n");
+        rv = clipboard_c2s_in_files(s, g_clip_c2s.data + 5, flist_size - 5);
+    }
+    else
+    {
+        LOG_DEVEL(LOG_LEVEL_ERROR,
+                  "clipboard_process_data_response_for_file: "
+                  "Unrecognised target");
+        rv = 1;
+    }
+
+    if (rv != 0 && g_clip_c2s.data != NULL)
+    {
+        g_clip_c2s.data[0] = '\0';
+    }
+
+    g_clip_c2s.total_bytes =
+        (g_clip_c2s.data == NULL) ? 0 : g_strlen(g_clip_c2s.data);
+    g_clip_c2s.read_bytes_done = g_clip_c2s.total_bytes;
+    clipboard_provide_selection_c2s(lxev, lxev->target);
+
+    return rv;
+}
+
 /*****************************************************************************/
 /* client to server */
 /* sent as a reply to CB_FORMAT_DATA_REQUEST; used to indicate whether
@@ -1227,26 +1288,8 @@ clipboard_process_data_response(struct stream *s, int clip_msg_status,
     }
     if (g_clip_c2s.xrdp_clip_type == XRDP_CB_FILE)
     {
-        g_free(g_clip_c2s.data);
-        g_clip_c2s.data = (char *)g_malloc(1024 * 1024, 1);
-        /* text/uri-list */
-        if (g_clip_c2s.type == g_file_atom1)
-        {
-            clipboard_c2s_in_files(s, g_clip_c2s.data);
-        }
-        /* x-special/gnome-copied-files */
-        else if (g_clip_c2s.type == g_file_atom2)
-        {
-            g_strcpy(g_clip_c2s.data, "copy\n");
-            clipboard_c2s_in_files(s, g_clip_c2s.data + 5);
-        }
-        else
-        {
-            LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_process_data_response: error");
-        }
-        g_clip_c2s.total_bytes = g_strlen(g_clip_c2s.data);
-        g_clip_c2s.read_bytes_done = g_clip_c2s.total_bytes;
-        clipboard_provide_selection_c2s(lxev, lxev->target);
+        clipboard_process_data_response_for_file(s, clip_msg_status,
+                clip_msg_len);
         return 0;
     }
     LOG_DEVEL(LOG_LEVEL_DEBUG, "clipboard_process_data_response: "

--- a/sesman/chansrv/clipboard.c
+++ b/sesman/chansrv/clipboard.c
@@ -1231,13 +1231,20 @@ clipboard_process_data_response_for_file(struct stream *s,
     /* text/uri-list */
     else if (g_clip_c2s.type == g_file_atom1)
     {
-        rv = clipboard_c2s_in_files(s, g_clip_c2s.data, flist_size);
+        rv = clipboard_c2s_in_files(s, g_clip_c2s.data, flist_size,
+                                    "file://");
     }
     /* x-special/gnome-copied-files */
     else if (g_clip_c2s.type == g_file_atom2)
     {
         g_strcpy(g_clip_c2s.data, "copy\n");
-        rv = clipboard_c2s_in_files(s, g_clip_c2s.data + 5, flist_size - 5);
+        rv = clipboard_c2s_in_files(s, g_clip_c2s.data + 5, flist_size - 5,
+                                    "file://");
+    }
+    else if ((g_clip_c2s.type == XA_STRING) ||
+             (g_clip_c2s.type == g_utf8_atom))
+    {
+        rv = clipboard_c2s_in_files(s, g_clip_c2s.data, flist_size, "");
     }
     else
     {
@@ -2211,11 +2218,24 @@ clipboard_event_selection_request(XEvent *xevent)
     }
     else if ((lxev->target == XA_STRING) || (lxev->target == g_utf8_atom))
     {
-        g_memcpy(&g_saved_selection_req_event, lxev,
-                 sizeof(g_saved_selection_req_event));
-        g_clip_c2s.type = lxev->target;
-        g_clip_c2s.xrdp_clip_type = XRDP_CB_TEXT;
-        clipboard_send_data_request(CB_FORMAT_UNICODETEXT);
+        if (clipboard_find_format_id(g_file_format_id) >= 0)
+        {
+            LOG_DEVEL(LOG_LEVEL_DEBUG, "clipboard_event_selection_request: "
+                      "text requested when files available");
+            g_memcpy(&g_saved_selection_req_event, lxev,
+                     sizeof(g_saved_selection_req_event));
+            g_clip_c2s.type = lxev->target;
+            g_clip_c2s.xrdp_clip_type = XRDP_CB_FILE;
+            clipboard_send_data_request(g_file_format_id);
+        }
+        else
+        {
+            g_memcpy(&g_saved_selection_req_event, lxev,
+                     sizeof(g_saved_selection_req_event));
+            g_clip_c2s.type = lxev->target;
+            g_clip_c2s.xrdp_clip_type = XRDP_CB_TEXT;
+            clipboard_send_data_request(CB_FORMAT_UNICODETEXT);
+        }
         return 0;
     }
     else if (lxev->target == g_image_bmp_atom)

--- a/sesman/chansrv/clipboard_file.c
+++ b/sesman/chansrv/clipboard_file.c
@@ -597,33 +597,41 @@ clipboard_c2s_in_file_info(struct stream *s, struct clip_file_desc *cfd)
 }
 
 /*****************************************************************************/
-/* See [MS-RDPECLIP] 2.2.5.2.3 */
 int
-clipboard_c2s_in_files(struct stream *s, char *file_list)
+clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size)
 {
-    int cItems;
+    int citems;
     int lindex;
     int str_len;
-    int file_count;
     struct clip_file_desc cfd;
     char *ptr;
+    char *last; /* Last writeable char in buffer */
+    int dropped_files = 0; /* # files we can't add to buffer */
+    const char *prefix = "file://";
 
-    if (!s_check_rem(s, 4))
+    if (file_list_size < 1)
     {
-        LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_c2s_in_files: parse error");
+        LOG(LOG_LEVEL_ERROR, "clipboard_c2s_in_files: No space in string");
         return 1;
     }
-    in_uint32_le(s, cItems);
-    if (cItems > 64 * 1024) /* sanity check */
+    if (!s_check_rem(s, 4))
     {
-        LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_c2s_in_files: error cItems %d too big", cItems);
+        LOG(LOG_LEVEL_ERROR, "clipboard_c2s_in_files: parse error");
+        return 1;
+    }
+    in_uint32_le(s, citems);
+    if (citems < 0 || citems > 64 * 1024) /* sanity check */
+    {
+        LOG(LOG_LEVEL_ERROR, "clipboard_c2s_in_files: "
+            "Bad number of files in list (%d)", citems);
         return 1;
     }
     xfuse_clear_clip_dir();
-    LOG_DEVEL(LOG_LEVEL_DEBUG, "clipboard_c2s_in_files: cItems %d", cItems);
-    file_count = 0;
+    LOG_DEVEL(LOG_LEVEL_DEBUG, "clipboard_c2s_in_files: cItems %d", citems);
     ptr = file_list;
-    for (lindex = 0; lindex < cItems; lindex++)
+    last = file_list + file_list_size - 1;
+
+    for (lindex = 0; lindex < citems; lindex++)
     {
         g_memset(&cfd, 0, sizeof(struct clip_file_desc));
         if (clipboard_c2s_in_file_info(s, &cfd) != 0)
@@ -633,37 +641,62 @@ clipboard_c2s_in_files(struct stream *s, char *file_list)
         if ((g_pos(cfd.cFileName, "\\") >= 0) ||
                 (cfd.fileAttributes & CB_FILE_ATTRIBUTE_DIRECTORY))
         {
-            LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_c2s_in_files: skipping directory not "
-                      "supported [%s]", cfd.cFileName);
+            LOG(LOG_LEVEL_WARNING, "clipboard_c2s_in_files: skipping "
+                "directory not supported [%s]", cfd.cFileName);
             continue;
         }
+
+        /* Have we already run out of room in the list? */
+        if (dropped_files > 0)
+        {
+            dropped_files += 1;
+            continue;
+        }
+
+        /* Room for this file? */
+        str_len = (ptr == file_list) ? 0 : 1; /* Delimiter */
+        str_len += g_strlen(prefix);
+        str_len += g_strlen(g_fuse_clipboard_path);
+        str_len += 1; /* '/' */
+        str_len += g_strlen(cfd.cFileName);
+        if (str_len > (last - ptr))
+        {
+            dropped_files += 1;
+            continue;
+        }
+
         if (xfuse_add_clip_dir_item(cfd.cFileName, 0, cfd.fileSizeLow, lindex) == -1)
         {
-            LOG_DEVEL(LOG_LEVEL_ERROR, "clipboard_c2s_in_files: failed to add clip dir item");
+            LOG(LOG_LEVEL_WARNING, "clipboard_c2s_in_files: "
+                "failed to add clip dir item %s", cfd.cFileName);
             continue;
         }
 
-        if (file_count > 0)
+        if (ptr > file_list)
         {
-            *ptr = '\n';
-            ptr++;
+            *ptr++ = '\n';
         }
-        file_count++;
 
-        g_strcpy(ptr, "file://");
-        ptr += 7;
+        str_len = g_strlen(prefix);
+        g_strcpy(ptr, prefix);
+        ptr += str_len;
 
         str_len = g_strlen(g_fuse_clipboard_path);
         g_strcpy(ptr, g_fuse_clipboard_path);
         ptr += str_len;
-
-        *ptr = '/';
-        ptr++;
+        *ptr++ = '/';
 
         str_len = g_strlen(cfd.cFileName);
         g_strcpy(ptr, cfd.cFileName);
         ptr += str_len;
     }
-    *ptr = 0;
+    *ptr = '\0';
+
+    if (dropped_files > 0)
+    {
+        LOG(LOG_LEVEL_WARNING, "clipboard_c2s_in_files: "
+            "Dropped %d files from the clip buffer due to insufficient space",
+            dropped_files);
+    }
     return 0;
 }

--- a/sesman/chansrv/clipboard_file.c
+++ b/sesman/chansrv/clipboard_file.c
@@ -598,7 +598,8 @@ clipboard_c2s_in_file_info(struct stream *s, struct clip_file_desc *cfd)
 
 /*****************************************************************************/
 int
-clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size)
+clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size,
+                       const char *fprefix)
 {
     int citems;
     int lindex;
@@ -607,7 +608,6 @@ clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size)
     char *ptr;
     char *last; /* Last writeable char in buffer */
     int dropped_files = 0; /* # files we can't add to buffer */
-    const char *prefix = "file://";
 
     if (file_list_size < 1)
     {
@@ -655,7 +655,7 @@ clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size)
 
         /* Room for this file? */
         str_len = (ptr == file_list) ? 0 : 1; /* Delimiter */
-        str_len += g_strlen(prefix);
+        str_len += g_strlen(fprefix); /* e.g. "file://" */
         str_len += g_strlen(g_fuse_clipboard_path);
         str_len += 1; /* '/' */
         str_len += g_strlen(cfd.cFileName);
@@ -677,8 +677,8 @@ clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size)
             *ptr++ = '\n';
         }
 
-        str_len = g_strlen(prefix);
-        g_strcpy(ptr, prefix);
+        str_len = g_strlen(fprefix);
+        g_strcpy(ptr, fprefix);
         ptr += str_len;
 
         str_len = g_strlen(g_fuse_clipboard_path);

--- a/sesman/chansrv/clipboard_file.h
+++ b/sesman/chansrv/clipboard_file.h
@@ -30,8 +30,24 @@ clipboard_process_file_request(struct stream *s, int clip_msg_status,
 int
 clipboard_process_file_response(struct stream *s, int clip_msg_status,
                                 int clip_msg_len);
+/**
+ * Process a CLIPRDR_FILELIST - see [MS-RDPECLIP] 2.2.5.2.3
+ *
+ * Files in the list are added to the xfs filesystem in the clipboard
+ * directory. The filenames names are added to the 'file_list' for the user.
+ * Files are prefixed with 'file://<clipboard_dir>/' and separated by '\n'.
+ *
+ * If the list is not big enough, whole filenames are omitted, and a warning
+ * message is logged. This is not an error.
+ *
+ * @param s Input stream containing CLIPRDR_FILELIST
+ * @param file_list Output buffer for filenames
+ * @param file_list_size Size of buffer, including space for '\0'.
+ *
+ * @return Zero for success.
+ */
 int
-clipboard_c2s_in_files(struct stream *s, char *file_list);
+clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size);
 
 int
 clipboard_request_file_size(int stream_id, int lindex);

--- a/sesman/chansrv/clipboard_file.h
+++ b/sesman/chansrv/clipboard_file.h
@@ -35,7 +35,7 @@ clipboard_process_file_response(struct stream *s, int clip_msg_status,
  *
  * Files in the list are added to the xfs filesystem in the clipboard
  * directory. The filenames names are added to the 'file_list' for the user.
- * Files are prefixed with 'file://<clipboard_dir>/' and separated by '\n'.
+ * Files are prefixed with '<fprefix><clipboard_dir>/' and separated by '\n'.
  *
  * If the list is not big enough, whole filenames are omitted, and a warning
  * message is logged. This is not an error.
@@ -43,11 +43,13 @@ clipboard_process_file_response(struct stream *s, int clip_msg_status,
  * @param s Input stream containing CLIPRDR_FILELIST
  * @param file_list Output buffer for filenames
  * @param file_list_size Size of buffer, including space for '\0'.
+ * @param fprefix Prefix for each file in the file list (e.g. "file://")
  *
  * @return Zero for success.
  */
 int
-clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size);
+clipboard_c2s_in_files(struct stream *s, char *file_list, int file_list_size,
+                       const char *fprefix);
 
 int
 clipboard_request_file_size(int stream_id, int lindex);

--- a/sesman/sesman.ini.in
+++ b/sesman/sesman.ini.in
@@ -125,6 +125,10 @@ FuseMountName=thinclient_drives
 FileUmask=077
 ; Can be used to disable FUSE functionality - see sesman.ini(5)
 #EnableFuseMount=false
+; Uncomment this line only if you are using GNOME 3 versions 3.29.92
+; and up, and you wish to cut-paste files between Nautilus and Windows. Do
+; not use this setting for GNOME 4, or other file managers
+#UseNautilus3FlistFormat=true
 
 [ChansrvLogging]
 ; Note: one log file is created per display and the LogFile config value 

